### DISCUSSION TITLE
Reference F Prime Mission (and hardware)

### DISCUSSION THREAD

**Discussion by SterlingPeet** [nasa/fprime#1403] [Ideas]
I have some personal interest in a reference mission, which goes all the way back to when I originally travelled to Pasadena to try and learn F Prime by immersion 😊

## 💡 Premise

I like learning within an appropriate context, and I'm sure other do too.  @timcanham has raised some very valid concerns about people getting hung up in the implementation of complicated constructs, and I **do** think that the Math Component Tutorial addresses that concern.  For me, once I am past the Math Tutorial, I find it most instructive to see some examples of how a realistic implementation works.  If the example is close, I can adapt it to my situation, if not, I can at least understand the approach to developing a solution.

## 🎈 Mission

To avoid EAR/ITAR issues, it is obvious that an educational robotic reference mission shouldn't be a cubesat 🛰️ (thanks government 🙄).  But even if that was feasible, it is hard to actually see and understand how everything might work in a controlled environment, because it is not in space.  The sensors and actuators are not ideal for the not-space demonstration.

That said, there are several things are more tractable on earth.  Here is my short list:

- 🤖 Mars Rover style crawler robot 
- 🦾 Desktop manipulator arm
- 🎈 Weather Balloon

The weather balloon is the one that I have put a little effort into myself, but it is not marginally ready for release.  I like it because it is easy for my students to directly transfer that knowledge and engineering over to a cubesat.  The parts are not very far away, and relatively easily adaptable (if you think ahead about generalizing designs).

That said, any mission that has a realistically complete implementation and ground data segment will be valuable as a reference.

## 📌 Considerations

Ah, the existential question: What are we doing here, and why?

I view this as an opportunity to provide some example implementation of things that are a more tangential to the core of F Prime.  Effort is not free, so it stands to reason that we want to get the most value return for the example mission.  Here are some points to consider:

- Parts used in the example are reasonably available (yea, its hard, theres no guarantees here)
- Re-use of existing example components (GPS maybe?) could boost overall value per unit work
- Common questions from the community are addressed (how do I connect a radio?)
- Power management is not largely considered in the main codebase
- An example of post-processing mission data could help with understanding of GDS pipeline, perhaps including a small telemetry database for data received during operation
- An example of a GDS plugin that works, but isn't in the GDS repository to understand out-of-tree modification therof

I am probably missing some, maybe there will be replies to this message that are helpful.

## 🚀 Some Specific Suggestions

These might be useful to consider.  Many of them are biased toward my needs, so consider that too.

### 💻 Computing Platform

My reference platform is set up for:

- BeagleBone Black - because they seem nearest to suitable for university cubesat missions
- Onion Omega 2+ - because it is low power, but not the easiest to cross compile
- Raspberry Pi - Already supported by F Prime, low effort index, easy camera add-on

### 📷 Camera

It seems like there have been a few requests relating to images, a raspberry pi with a camera might be a relatively low effort integration.  It would also provide a convenient example for dealing with file downlinks.

### 🦾 Actuators

Hobby servos are probably the easy thing here.  Should already exist.  For a weather balloon, the servo could actuate a cutdown mechanism, or it could tilt a camera.

### 📡 Comms/Radio

This is probably the thing with the biggest value in having an example.  Using WiFi is convenient, but it tends not to yield a realistic example of how anyone really needs to use F Prime.  Whatever is used should be relatively standardized and easy to get.  Avoiding special licenses probably pushes toward ISM radios, and likely 900 MHz unless an explicitly amateur radio example is desired.

There are basically 2 ultra common radio varieties that have proliferated in the hobby circles that I have noticed, they are:

- LoRa radios based on the RFM95 type (which is really a Semtech SX1276 under the hood)
- XBee radios from Digi

Both have good documentation and tons of community code supporting them.  The RFM9X radios are lower level devices that require register manipulation, although there are good libraries available for many platforms to do that.  The XBees are higher level devices that can be operated as UART passthrough, configured with an AT command set, or operated at a lower level with SPI commands; depending on the firmware flashed and features desired.

## 🎥 Epilogue

Such a reference mission would greatly benefit the community, so I think it is a well deserved effort.  It can be very hard for people to imagine all the ways F Prime can be used or implemented, and I think that is partly due to F Prime's flexibility.  Building full featured remote robotic platforms turns out to be complicated.   On the other hand, a working example mission removed the need for significant imagination about **how** it is possible to implement a mission, and that should boost adoption for F Prime.

@LeStarch hopefully some of these comments are useful to you.

Thanks,

@SterlingPeet

**Comment by LeStarch** [comment]
This helps a lot.  Thanks!

**Comment by LeStarch** [comment]
See also: #1365
